source/synthDrivers/_sonic.py

```diff
@@ -0,0 +1,203 @@
+	sonicLib.sonicDestroyStream.argtypes = [SonicStreamP]
+	sonicLib.sonicWriteFloatToStream.restype = c_int
+	sonicLib.sonicWriteFloatToStream.argtypes = [SonicStreamP, c_void_p, c_int]
+	sonicLib.sonicWriteShortToStream.restype = c_int
+	sonicLib.sonicWriteShortToStream.argtypes = [SonicStreamP, c_void_p, c_int]
+	sonicLib.sonicWriteUnsignedCharToStream.restype = c_int
+	sonicLib.sonicWriteUnsignedCharToStream.argtypes = [SonicStreamP, c_void_p, c_int]
+	sonicLib.sonicReadFloatFromStream.restype = c_int
+	sonicLib.sonicReadFloatFromStream.argtypes = [SonicStreamP, c_void_p, c_int]
+	sonicLib.sonicReadShortFromStream.restype = c_int
+	sonicLib.sonicReadShortFromStream.argtypes = [SonicStreamP, c_void_p, c_int]
+	sonicLib.sonicReadUnsignedCharFromStream.restype = c_int
+	sonicLib.sonicReadUnsignedCharFromStream.argtypes = [SonicStreamP, c_void_p, c_int]
+	sonicLib.sonicFlushStream.restype = c_int
+	sonicLib.sonicFlushStream.argtypes = [SonicStreamP]
+	sonicLib.sonicSamplesAvailable.restype = c_int
+	sonicLib.sonicSamplesAvailable.argtypes = [SonicStreamP]
+	sonicLib.sonicGetSpeed.restype = c_float
+	sonicLib.sonicGetSpeed.argtypes = [SonicStreamP]
+	sonicLib.sonicSetSpeed.restype = None
+	sonicLib.sonicSetSpeed.argtypes = [SonicStreamP, c_float]
+	sonicLib.sonicGetPitch.restype = c_float
+	sonicLib.sonicGetPitch.argtypes = [SonicStreamP]
+	sonicLib.sonicSetPitch.restype = None
+	sonicLib.sonicSetPitch.argtypes = [SonicStreamP, c_float]
+	sonicLib.sonicGetRate.restype = c_float
+	sonicLib.sonicGetRate.argtypes = [SonicStreamP]
+	sonicLib.sonicSetRate.restype = None
+	sonicLib.sonicSetRate.argtypes = [SonicStreamP, c_float]
+	sonicLib.sonicGetVolume.restype = c_float
+	sonicLib.sonicGetVolume.argtypes = [SonicStreamP]
+	sonicLib.sonicSetVolume.restype = None
+	sonicLib.sonicSetVolume.argtypes = [SonicStreamP, c_float]
+	sonicLib.sonicGetQuality.restype = c_int
+	sonicLib.sonicGetQuality.argtypes = [SonicStreamP]
+	sonicLib.sonicSetQuality.restype = None
+	sonicLib.sonicSetQuality.argtypes = [SonicStreamP, c_int]
+	sonicLib.sonicGetSampleRate.restype = c_int
+	sonicLib.sonicGetSampleRate.argtypes = [SonicStreamP]
+	sonicLib.sonicSetSampleRate.restype = None
+	sonicLib.sonicSetSampleRate.argtypes = [SonicStreamP, c_int]
+	sonicLib.sonicGetNumChannels.restype = c_int
+	sonicLib.sonicGetNumChannels.argtypes = [SonicStreamP]
+	sonicLib.sonicSetNumChannels.restype = None
+	sonicLib.sonicSetNumChannels.argtypes = [SonicStreamP, c_int]
+
+
+class SonicStream:
+	"""
+	Audio stream that wraps the Sonic library to process audio,
+	which is optimised for speeding up speech by high factors.
+	Audio data are stored internally as 16-bit integers.
+	"""
+
+	def __init__(self, sampleRate: int, channels: int):
+		self.stream: SonicStreamP = sonicLib.sonicCreateStream(sampleRate, channels)
+		if not self.stream:
+			raise MemoryError()
+
+	def __del__(self):
+		sonicLib.sonicDestroyStream(self.stream)
+
+	def writeFloat(self, data: c_void_p, numSamples: int) -> None:
```

Comment:
Can these write functions get more specific types? Something like
```suggestion
	def writeFloat(self, data: c_float_p, numSamples: int) -> None:
```
And at the top of the file:
```py
from typing import TYPE_CHECKING

if TYPE_CHECKING:
	from ctypes import _Pointer
	c_float_p = _Pointer[c_float]
else:
	c_float_p = POINTER(c_float)
```

Reply:
In fact, in `sapi5.py`, a pointer to `c_ubyte` is passed into `writeShort`, not a pointer to `c_short` as it's supposed to be.

But sonic accepts data in whole samples rather than bytes, so maybe it's better to make that explicit.